Issue #7066: correct RequireThis for expression cases

strkkk opened the pull request:
#7066 

Report: https://strkkk.github.io/7066_2/index.html

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheck.java:
```diff
@@ -615,7 +631,7 @@ private AbstractFrame getClassFrameWhereViolationIsFound(DetailAST ast) {
         if (variableDeclarationFrameType == FrameType.CLASS_FRAME
                 && !validateOnlyOverlapping
-                && prevSibling == null
+                && (prevSibling == null || EXPR_TOKENS.contains(ast.getParent().getType()))
```
Curious, what changes occur when you remove `prevSibling == null`? Why don't we just look at all expressions?

strkkk replied:
it breaks cases like
```
public int doSideEffect() {
       return bar;
}
```

rnveach replied:
When you say it breaks cases like that, do you mean removing it from the old code? The new code makes sense since you are checking for specific tokens, but why can't we do all expressions?

strkkk replied:
this check (prevSibling == null) relates not only to expressions.

rnveach replied:
> it breaks cases like
> return bar;

Bar is still part of an expression, so I don't see how removing `prevSibling == null` breaks it.

````
    |  `--SLIST -> { [2:26]
    |      |--LITERAL_RETURN -> return [3:7]
    |      |  |--EXPR -> EXPR [3:14]
    |      |  |  `--IDENT -> bar [3:14]
    |      |  `--SEMI -> ; [3:17]
    |      `--RCURLY -> } [4:0]
````

rnveach replied:
I removed the entire line, `&& (prevSibling == null || EXPR_TOKENS.contains(ast.getParent().getType()))`, and false positives came up like:
> `this.booleanField = true;`
> `if (this.ex != null) {`
> `Object plus = "" + this.id;`

It looks like it is requesting everything have a double `this`. It seems to me this line just needs to check there isn't already a `this` but it would have to be examined more.

strkkk replied:
looks like a good idea, I didn't notice it. Will try it later

strkkk replied:
your suggestion worked out well, thanks. I corrected code.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheck.java:
```diff
@@ -267,6 +268,21 @@
+            TokenTypes.LT,
+            TokenTypes.GT,
+            TokenTypes.LE,
+            TokenTypes.GE
```
What about left / right shift?

strkkk replied:
Yeap, I forgot some of the operations (not only these)
I added missing ones and tests

rnveach replied:
This is why I didn't like a hardcoded list. It will be easy to miss some operators.
Why can't we check all expressions?

strkkk replied:
How, for example?

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheck.java:
```diff
@@ -267,6 +268,21 @@
+            TokenTypes.LT,
+            TokenTypes.GT,
+            TokenTypes.LE,
+            TokenTypes.GE
```
Can we sort/group these somehow? Either Alphabetical or by conditions / math operators?

strkkk replied:
I changed order to alphabetical

rnveach commented:
https://strkkk.github.io/7066_2/Orekit/index.html#A13
https://strkkk.github.io/7066_2/Orekit/xref/checkstyle/contribution/checkstyle-tester/repositories/Orekit/src/main/java/org/orekit/bodies/Ellipse.java.html#L206
I found this area to be an interesting case.

`a` has a new violation on this line, but not the next line. `a` was already being violated there. The side of the expression matters.
This seems connected to my question at https://github.com/checkstyle/checkstyle/pull/7069#discussion_r324855780

strkkk commented:
> `a` has a new violation on this line, but not the next line. `a` was already being violated there. The side of the expression matters in the old code.

There is no puzzle. Code was checking if previous sibling is null, and for first member of the expression it is true, but false for second. Now it is corrected.

rnveach commented:
> There is no puzzle.

I was more pointing it out as your new test cases only have it 1 way and not the other. I wasn't sure how much the other way was being tested already.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheck.java:
```diff
@@ -615,7 +639,7 @@ private AbstractFrame getClassFrameWhereViolationIsFound(DetailAST ast) {
         if (variableDeclarationFrameType == FrameType.CLASS_FRAME
                 && !validateOnlyOverlapping
-                && prevSibling == null
+                && (prevSibling == null || EXPR_TOKENS.contains(ast.getParent().getType()))
```
There is similar code 16 lines down. Does that need to be changed too?

strkkk replied:
I tried to remove it and it broke this test case:

```
   private String field1;
    void foo8(Long field1) {
        field1 += field1; // violation
    }
```
with config validateOnlyOverlapping = false and other options = default

I am puzzled how it should be checked because it have 3 options (add this to both or add to only one) and it is not clear what is correct.

Can you clarify it?

rnveach replied:
> I tried to remove it 

Are you talking about removing the code 16 lines down? Why remove it when this item was asking about updating it to be insync with the changes we are making here?

strkkk replied:
@rnveach nevemind. The thing is that this condition (prevSibling == null) in condition below is responsible for test case
```
private String field1;
    void foo8(Long field1) {
        field1 += field1; // violation
    }
```
Can you clarify for this case, what should be a violation? It is not clear for me.

rnveach replied:
> The thing is that this condition (prevSibling == null) in condition below is responsible for test case

Just so we are clear, I am not asking to remove that condition for this PR.

> Can you clarify for this case, what should be a violation? It is not clear for me.

@romani ping. Can you clarify this line.
I want to say that if the same name is used in both places and the name represents a field and a non-field, we request the the name on the left of the assignment have a `this.` as it most likely should represent the field and the non-field is blocking the correct assignment.

strkkk replied:
@rnveach I created a diff report for this condition.
Base branch is this one, patch branch contains same check for condition below
https://strkkk.github.io/7066_3/index.html

It looks like adding this condition will cause a lot of false-positives, e.g.
https://strkkk.github.io/7066_3/lombok-ast/xref/checkstyle/contribution/checkstyle-tester/repositories/lombok-ast/src/ecjTransformer/lombok/ast/ecj/EcjTreeConverter.java.html#L230

rnveach replied:
I agree, it looks like changing that code is a bad idea.
So we won't do this. Thanks for confirming.

romani replied:
I running late .... sorry.
Please read documentation of this Check on such cases. I agree it is gray area of this Check.
Please try to analyze what is now and what it should be.

I will be back after I unblock eclipse-cs project, sorry.

romani commented:
@strkkk , ping

strkkk commented:
@romani I will not be able to work on it for next 3 weeks.

romani commented:
No problem, thanks a lot for notification. Take your time.

romani commented:
@strkkk , ping, is the any resources to finish?

strkkk commented:
@romani Will be next week

El mié., 23 oct. 2019 a las 18:28, Roman Ivanov (<notifications@github.com>)
escribió:

> @strkkk <https://github.com/strkkk> , ping, is the any resources to
> finish?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/checkstyle/checkstyle/pull/7069?email_source=notifications&email_token=ACD5F2SYTTL2Q74FABARHKDQQBUSZA5CNFSM4IXEIELKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECB2YJI#issuecomment-545500197>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACD5F2XXCPCAQKWXNDVP6SLQQBUSZANCNFSM4IXEIELA>
> .
>

strkkk commented:
"awaiting changes" label can be removed :)

rnveach commented:
I'm still waiting on https://github.com/checkstyle/checkstyle/pull/7069#discussion_r325431018 to be done :)

strkkk commented:
Ok, then I got it wrong.
I thought this means something like "PR is frozen for some time"

Regarding your comment - lets wait for clarification.

romani commented:
In my vision of this label, we waited for some actions from author.
If decision is blocked by me, please remove label and assign PR to me, I will be sure that this PR require response from me

rnveach commented:
@strkkk clarification of that example doesn't have anything to do with the change I am asking for.
2 if statements that I pointed to have very similar conditions. You changed one of them to add a brand new condition. I am just looking for if the other if statement should add a similar condition.

strkkk commented:
@rnveach Yes, I understand. The thing is that I do not understand what second condition is responsible for, I do not see cases when adding it will make a difference.

When I removed it just to see why it is there, test failed with the case that I mentioned. I have no idea what should be correct there and is there any need for `this` at all. Without it I can't say should we add there same condition as one I already added or not.

rnveach commented:
If your unsure what effect adding the new condition to the other location will have, you can always add it and run regression between that version and this branch to see if any differences show up. If differences show up for that 1 little change, then we can review them to see if they are valid or not. If no differences show up, then we can ignore adding it as we can produce any need for it.

strkkk commented:
@romani ping

romani commented:
@strkkk , please pull latest version of checkstyle-tester, report rendering was updated few month ago 
https://github.com/checkstyle/contribution/issues/362
